keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +333,54 @@ or later. The dual-stack feature requires no changes to this API.
 
 ### Services
+Services depend on `--services-cluster-ip-range` for `ClusterIP` assignment. A dualstack cluster can have this flag configured as `--services-cluster-ip-rang=<cidr>,<cidr>` or `--services-cluster-ip-rang=<cidr>`. The following rules apply:
+
+1. The maximum allowed number of cidrs in this flag is two. Attempting to use more than two cidrs will generate failures during startup. 
+2. When configured with two cidrs then the values are expected to have different ip families. e.g `--services-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>` or `--services-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`. Attempting to configure a cluster with two similar cidrs will generate failures during startup.
+3. A dualstack cluster can have a single cidr in this flag. That means the cluster will be to egress traffic on both ip families, but will *not* be ingress traffic on both ip families.
+4. When upgrading a cluster to dualstack, the first element of the `--services-cluster-ip-range` flag MUST NOT change IP families.  As an example, an IPv4 cluster can safely be upgraded to use `--services-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>`, but can not safely be changed to `--services-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`.
+
 
 #### Type NodePort, LoadBalancer, ClusterIP
-Services will only be assigned an address from a single address family (either IPv4 or IPv6). The address family for the Service’s cluster IP is determined by setting the field, `.spec.ipFamily`, on that Service. You can only set this field when creating a new Service. Setting the `.spec.ipFamily` field is optional and should only be used if you plan to enable IPv4 and IPv6 Services and Ingresses on your cluster. The configuration of this field not a requirement for egress traffic. The default address family for your cluster is the address family of the first service cluster IP range configured via the `--service-cluster-ip-range` flag to the kube-controller-manager. The `.spec.ipFamily` field can be set to either IPv4 or IPv6.
+Services will only be assigned an address from a single address family (either IPv4 or IPv6). This assignment is user controlled by setting the value of `.spec.ipFamily` on that Service. Setting the `.spec.ipFamily` field is optional and should only be used when there are multiple entries configured in `--services-cluster-ip-range` flag. The `.spec.ipFamily` field allows mutability under the same conditions `.spec.ClusterIP` field allows it. Another field will be introduced in Service is `.status.ipFamilies` field. The value of this field is an array of `ipFamily` where `.status.ipFamilies[0]` is used by endpoints controller to select the appropriate endpoints. The value of this field is set according to the following rules:
```

Comment:
just a comment as a reminder about this bug https://github.com/kubernetes/kubernetes/issues/90101#issuecomment-615347456 , where both IP Families can not share the same NodePort

Reply:
I don't think this is relevant to a `kep`. No matter how we solve for that bug, the kep won't need to change according to the outcome.

Reply:
my question is if each IP family will have its nodeport space independent of each other, or both families will share the same pool of nodeports.
I have preference for the former, but don't know if everybody agrees, and I think that is part of the architectural design ... I don't have clear if is easy to implement a "SecondaryNodePortAllocator"

Reply:
Let's keep #90101 discussion in #90101 :)